File: labthree/LabThree.m
#import "LabThree.h"

double power(double a, int n) {
    if (n == 0)
        return 1;
    
    double t = power(a, n / 2);
    
    if (n % 2 == 0)
        return t * t;
    else
        if (n > 0)
            return a * t * t;
        else
            return (t * t) / a;
}

@implementation LabThree

+ (void)mergeSort:(NSMutableArray *)list {
    NSMutableArray *sideList = [NSMutableArray arrayWithArray:list];

    for (int i = 1; i < list.count; i *= 2) {
        for (int j = 0; j < list.count; j += i * 2)
            [LabThree mergeSortWithList:list
                              leftIndex:j
                             rightIndex:min(i + j, (int) list.count)
                                    tip:min(j + 2 * i, (int) list.count)
                               sideList:sideList];
        [list setArray:sideList];
    }
}

+ (void)mergeSortWithList:(NSMutableArray *)list
                leftIndex:(int)leftIndex
               rightIndex:(int)rightIndex
                      tip:(int)tip
                 sideList:(NSMutableArray *)sideList {

    int i = leftIndex, j = rightIndex;

    for (int k = leftIndex; k < tip; k++)
        if (i < rightIndex && (j >= tip || list[i] <= list[j]))
            sideList[k] = list[i++];
        else
            sideList[k] = list[j++];
}

+ (int)binarySearchList:(NSMutableArray *)list toFindElement:(id)element {
    int lowestIndex = 0, highestIndex = (int) list.count - 1;

    while (lowestIndex <= highestIndex) {
        int indexInBetween = lowestIndex + (highestIndex - lowestIndex) / 2;
        if (element < list[indexInBetween])
            highestIndex = indexInBetween - 1;
        else if (element > list[indexInBetween])
            lowestIndex = indexInBetween + 1;
        else
            return indexInBetween;
    }

    return -1;
}

@end

File: Lab 3/labthree.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

def power(base, exponent):
    if exponent == 0:
        return 1

    total = power(base, exponent / 2)

    if exponent % 2 == 0:
        return total ** 2
    else:
        return base * total ** 2 if exponent > 0 else (total ** 2) / base


def merge_sort(array):
    array_length = len(array)
    side_array = [0 for i in range(array_length)]

    i = 1
    while i < array_length:
        j = 0
        while j < array_length:
            merge(array, j, min(i + j, array_length),
                  min(j + 2 * i, array_length), side_array)
            j += i * 2
        array = side_array[:]
        i *= 2

    return array


def merge(array, left_index, right_index, tip, side_array):
    i, j = left_index, right_index
    for k in range(left_index, tip):
        if i < right_index and (j >= tip or array[i] <= array[j]):
            side_array[k] = array[i]
            i += 1
        else:
            side_array[k] = array[j]
            j += 1


def binary_search(array, element):
    lowest_index, highest_index = 0, len(array) - 1

    while lowest_index <= highest_index:
        index_in_between = lowest_index + (highest_index - lowest_index) / 2
        if element < array[index_in_between]:
            highest_index = index_in_between - 1
        elif element > array[index_in_between]:
            lowest_index = index_in_between + 1
        else:
            return index_in_between

    return -1


def main():
    array = [5, 1, 3, 6, 21, 10, 35, 44, 12]
    print "Unsorder array: %r" % array
    array = merge_sort(array)
    print "Sorted array: %r" % array

    print "Found 21 at index: %d" % binary_search(array, 21)

    print "5^5 = %.0f" % power(5, 5)
    print "20^21 = %.0f" % power(20, 21)


if __name__ == "__main__":
    main()
